Cython/Compiler/ExprNodes.py

```diff
@@ -2035,13 +2044,25 @@ def declare_from_annotation(self, env, as_target=False):
                           "'%s' cannot be specialized since its type is not a fused argument to this function" %
                           self.name)
                     atype = error_type
+            visibility = 'private'
+            if 'dataclasses.dataclass' in env.directives:
+                # handle "frozen" directive - full inspection of the dataclass directives happens
+                # in Dataclass.py
+                frozen_directive = None
+                dataclass_directive = env.directives['dataclasses.dataclass']
+                if dataclass_directive:
+                    dataclass_directive_kwds = dataclass_directive[1]
+                    frozen_directive = dataclass_directive_kwds.get('frozen', None)
+                is_frozen = frozen_directive and frozen_directive.is_literal and frozen_directive.value
```

Comment:
There's also `constant_result` after constant folding, with `has_constant_result()` as a check. Not sure if that would be better here.

Reply:
It seems to be too early to use `has_constant_result` so I think I think this should remain as is.